keps/sig-node/5328-node-capabilities/README.md

```diff
@@ -0,0 +1,859 @@
+
+#### Deprecation
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+-->
+
+**Alpha:**
+*   The `NodeCapabilities` feature gate is implemented and disabled by default.
+*   The Kubelet correctly discovers and reports capabilities to the `node.status.capabilities` map when the feature gate is enabled.
+*   The API server correctly serves and validates the `capabilities` field when the feature gate is enabled.
+*   The `kube-scheduler` and `admission-controller` plugins are introduced.
+*   All unit tests outlined in the Test Plan are implemented and verified.
+
+
+### Upgrade / Downgrade Strategy
+
+#### Upgrade
+
+Users can enable the Node Capabilities feature gate after upgrading to a Kubernetes version that supports it. For the feature to be fully effective, both the control plane and Kubelets must be upgraded to this version. Existing workload should not after the upgrade. The [Version Skew Strategy](#version-skew-strategy) section details behavior when only the control plane is upgraded while nodes are still running older Kubelet versions without the feature.
+
+
+#### Downgrade
+
+Downgrading both kubelet and control plane components to a version without the Node Capabilities feature means subsequent scheduling decisions and API request validations will no longer utilize these capabilities. The [Version Skew Strategy](#version-skew-strategy) section details behavior when only the kubelet is downgraded to a version lacking the feature.
+
+### Version Skew Strategy
+
+Scenario 1: Node Capabilities enabled in the Control Plane (v1.X or feature gate ON), not enabled in kubelet  (v1.X-1 or feature gate OFF). 
+
+*   This presents a challenge when existing features begin to be reported through the new Node Capabilities mechanism. Few possible options to manage this are
+    1. Implement phased capability reliance in the Control Plane. The newer (v1.X) control plane would initially avoid basing decisions based on the new capability signals, and would begin to do so past the allowed version skew (v1. X + version skew).
+    2. Backfill capability information on older versions.
+
+Scenario 2 :  kube-scheduler is looking for certain kubelet capability (a new feature added in v1.X) but nodes are running older kubelet versions ( v1.X-2) that do not have the capability.
+
+*   This is a valid use case that Node Capabilities is trying to address. The pod that relies on the feature being enabled should remain pending with a valid status message. 
+*   This prevents pods from failing or running with degraded functionality on incompatible nodes, which can happen today if feature-specific node labels and selectors are not properly applied.
+
```

Comment:
couldn't we use this feature? if the node.Status has `feature.kubernetes.io/nodeCapabilities`, then the scheduler will prefer those nodes for special features, but consider ones with out that (or any node capabilities). admission plugin won't validate unless feature.kubernetes.io/nodeCapabilities is present

Reply:
@haircommander I don't follow. Are you suggesting the idea to have "soft requirements" for capabilities? I think we've scoped it to only cover hard-requirements for now. If there's a fallback mode where running without the capability is acceptable, then I don't think capabilities as described here is a good fit.

@pravk03 this might a good one to list under the possible future enhancements section.

Reply:
@haircommander Is your suggestion for Scenario 1 ?. i,e to differenciate between nodes where feature flag is disabled and nodes which do not have a feature specific capability to report. In that case, I think this is a valid suggestion.